File: backend/src/outreach/schemas/leads.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

LeadField = Literal["email", "first_name", "last_name", "phone", "linkedin_url", "company", "title"]


class LeadCreate(BaseModel):
    email: str | None = None
    first_name: str | None = None
    last_name: str | None = None
    phone: str | None = None
    linkedin_url: str | None = None
    company: str | None = None
    title: str | None = None


class LatestReplyOut(BaseModel):
    event_id: int
    occurred_at: datetime
    body: str | None
    channel: str = "email"
    sentiment_label: str | None
    sentiment_confidence: float | None
    sentiment_reasoning: str | None


class LeadOut(BaseModel):
    id: int
    email: str | None
    phone: str | None
    linkedin_url: str | None
    first_name: str | None
    last_name: str | None
    company: str | None
    title: str | None
    source: str | None
    created_at: datetime
    updated_at: datetime
    latest_reply: LatestReplyOut | None = None
    channel_replies: dict[str, LatestReplyOut] = Field(default_factory=dict)


class LeadListResponse(BaseModel):
    items: list[LeadOut]
    total: int
    limit: int
    offset: int


class UploadPreviewResponse(BaseModel):
    token: str
    row_count: int
    columns: list[str]
    sample_rows: list[dict]
    suggested_mapping: dict[LeadField, str | None]


class UploadCommitRequest(BaseModel):
    token: str
    mapping: dict[LeadField, str | None]
    source: str = Field(default="csv", max_length=40)


class UploadCommitResponse(BaseModel):
    inserted: int
    updated: int
    merged_within_upload: int = 0
    skipped_no_identity: int
    skipped_invalid: int

File: frontend/src/api/leads.ts
import axios from "axios";

import { api } from "./client";

export type LeadField = "email" | "first_name" | "last_name" | "phone" | "linkedin_url" | "company" | "title";

export interface LatestReply {
  event_id: number;
  occurred_at: string;
  body: string | null;
  channel: string;
  sentiment_label: string | null;
  sentiment_confidence: number | null;
  sentiment_reasoning: string | null;
}

export interface LeadOut {
  id: number;
  email: string | null;
  phone: string | null;
  linkedin_url: string | null;
  first_name: string | null;
  last_name: string | null;
  company: string | null;
  title: string | null;
  source: string | null;
  created_at: string;
  updated_at: string;
  latest_reply: LatestReply | null;
  channel_replies: Record<string, LatestReply>;
}

export interface LeadListResponse {
  items: LeadOut[];
  total: number;
  limit: number;
  offset: number;
}

export interface UploadPreviewResponse {
  token: string;
  row_count: number;
  columns: string[];
  sample_rows: Record<string, string>[];
  suggested_mapping: Record<LeadField, string | null>;
}

export interface UploadCommitResponse {
  inserted: number;
  updated: number;
  merged_within_upload: number;
  skipped_no_identity: number;
  skipped_invalid: number;
}

export async function previewUpload(file: File): Promise<UploadPreviewResponse> {
  const fd = new FormData();
  fd.append("file", file);
  const r = await api.post<UploadPreviewResponse>("/leads/upload/preview", fd, {
    headers: { "Content-Type": "multipart/form-data" },
  });
  return r.data;
}

export async function commitUpload(body: {
  token: string;
  mapping: Record<LeadField, string | null>;
  source?: string;
}): Promise<UploadCommitResponse> {
  const r = await api.post<UploadCommitResponse>("/leads/upload/commit", body);
  return r.data;
}

export async function listLeads(params: { search?: string; limit?: number; offset?: number } = {}): Promise<LeadListResponse> {
  const r = await api.get<LeadListResponse>("/leads", { params });
  return r.data;
}

export async function createLead(dto: Partial<Omit<LeadOut, "id" | "created_at" | "updated_at" | "source">>): Promise<LeadOut> {
  const r = await api.post<LeadOut>("/leads", dto);
  return r.data;
}

export async function updateLead(id: number, dto: Partial<Omit<LeadOut, "id" | "created_at" | "updated_at" | "source">>): Promise<LeadOut> {
  const r = await api.patch<LeadOut>(`/leads/${id}`, dto);
  return r.data;
}

export async function deleteLead(id: number): Promise<void> {
  await api.delete(`/leads/${id}`);
}

export function formatAxiosError(e: unknown): string {
  if (axios.isAxiosError(e)) {
    return e.response?.data?.detail ? `${e.response.status}: ${JSON.stringify(e.response.data.detail)}` : e.message;
  }
  return String(e);
}
